File: app/src/main/java/org/codroid/editor/widget/CodroidEditor.java
/*
 *     Copyright (c) 2021 Zachary. All rights reserved.
 *
 *     This file is part of Codroid.
 *
 *     Codroid is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     Codroid is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with Codroid.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.codroid.editor.widget;

import static org.codroid.editor.UIUtilsKt.dip2px;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupWindow;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatEditText;

import org.codroid.editor.R;
import org.codroid.interfaces.addon.AddonManager;
import org.codroid.interfaces.appearance.AppearanceProperty;
import org.codroid.interfaces.appearance.Part;
import org.codroid.interfaces.appearance.editor.WrappedSpannable;
import org.codroid.interfaces.appearance.parts.EditorPart;
import org.codroid.interfaces.evnet.EventCenter;
import org.codroid.interfaces.evnet.editor.SelectionChangedEvent;
import org.codroid.interfaces.evnet.editor.TextChangedEvent;

import java.util.Optional;

/**
 * CodroidEditor is a core part of this project.
 * It contains the function of highlight, autocomplete and etc.
 */

public class CodroidEditor extends AppCompatEditText {

    private Configure mConfigure;
    private WrappedSpannable mWrappedSpannable;

    private Paint lineNumberPaint, backPaint;

    private Rect lineNumberBackRect;

    // This is suggestion window
    private PopupWindow mPopupWindow;

    // The part of editor that defines the appearances
    private Optional<Part> part;

    private int paddingLeft;
    private int gap;

    // return true if editor is drawn at first time.
    private boolean isDrawn = false;

    public CodroidEditor(@NonNull Context context) {
        super(context);
        init();
    }

    public CodroidEditor(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public CodroidEditor(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        if (mConfigure == null) mConfigure = new Configure();

        if (lineNumberPaint == null && mConfigure.isShowLineNumber()) {
            lineNumberPaint = new Paint();
            lineNumberPaint.setStyle(Paint.Style.FILL);
            lineNumberPaint.setTextSize(getTextSize());
            lineNumberPaint.setColor(mConfigure.getLineNumberColor());

            backPaint = new Paint();
            backPaint.setStyle(Paint.Style.FILL);
            backPaint.setColor(mConfigure.getLineNumberBackColor());

            part = AddonManager.get().appearancePart(AppearanceProperty.PartEnum.EDITOR);

            part.ifPresent(p -> p.findColor(EditorPart.Attribute.BACKGROUND,
                    value -> setBackgroundColor(value.toArgb())));
            lineNumberBackRect = new Rect();
        }
        mWrappedSpannable = new WrappedSpannable(getText());

    }

    private void initSuggestionWindow() {
        if (mPopupWindow == null) {
            View view = View.inflate(getContext(), R.layout.suggestion_window, null );
            mPopupWindow = new PopupWindow(view, ViewGroup.LayoutParams.MATCH_PARENT, 300);
        }
    }

    @Override
    public void setText(CharSequence text, BufferType type) {
        super.setText(text, type);
    }

    @Override
    public void layout(int l, int t, int r, int b) {
        super.layout(l, t, r, b);
        gap = (int) dip2px(getContext(), 4);
        paddingLeft = (int) lineNumberPaint.measureText(String.valueOf(getLineCount())) + gap * 2;
        setPadding(paddingLeft + gap, 0, 0, 0);

    }

    public Configure getConfigure() {
        return mConfigure;
    }

    public void setConfigure(Configure mConfigure) {
        this.mConfigure = mConfigure;
    }


    @Override
    protected void onSelectionChanged(int selStart, int selEnd) {
        super.onSelectionChanged(selStart, selEnd);
        AddonManager.get().eventCenter()
                .<SelectionChangedEvent>execute(EventCenter.EventsEnum.EDITOR_SELECTION_CHANGED)
                .forEach(it -> {
                    try {
                        it.onSelectionChanged(mWrappedSpannable.update(getText()), selStart, selEnd);
                    } catch (Exception e) {
                        AddonManager.get().getLogger().e("Event editor_selection_changed executes failed with: " + e);
                    }
                });
    }

    @Override
    protected void onTextChanged(CharSequence text, int start, int lengthBefore, int lengthAfter) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter);
        AddonManager.get().eventCenter()
                .<TextChangedEvent>execute(EventCenter.EventsEnum.EDITOR_TEXT_CHANGED)
                .forEach(it -> {
                    try {
                        it.onTextChanged(mWrappedSpannable.update(getText()), start, lengthBefore, lengthAfter);
                    } catch (Exception e) {
                        AddonManager.get().getLogger().e("Event editor_text_changed executes failed with: " + e);
                    }
                });
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (!isDrawn) isDrawn = true;
        //Drawing line number
        if (mConfigure.isShowLineNumber()) {
            lineNumberBackRect.set(0, 0, paddingLeft, getHeight());
            canvas.drawRect(lineNumberBackRect, backPaint);
            for (int i = 0; i < getLineCount(); i++) {
                int baseLine = getLineBounds(i, null);
                canvas.drawText(String.valueOf(i + 1), gap, baseLine, lineNumberPaint);
            }
        }
    }

    public void showSuggestionWindow() {
        initSuggestionWindow();
        mPopupWindow.dismiss();
        mPopupWindow.showAtLocation(this, Gravity.TOP | Gravity.START, 0,
                getLineBounds(getCurrentCursorLine() - 1, null) + mPopupWindow.getHeight() + getLineHeight() + 30);
    }

    private int getCurrentCursorLine() {
        int selectionStart = getSelectionStart();
        if (selectionStart != -1) {
            return getLayout().getLineForOffset(selectionStart) + 1;
        }
        return -1;
    }


    public class Configure {
        private boolean lineNumber = true;
        private @ColorInt
        int lineNumberColor = Color.BLACK;
        private @ColorInt
        int lineNumberBackColor = Color.GRAY;

        public boolean isShowLineNumber() {
            return lineNumber;
        }

        public Configure showLineNumber(boolean lineNumber) {
            this.lineNumber = lineNumber;
            return this;
        }

        public @ColorInt
        int getLineNumberColor() {
            return lineNumberColor;
        }

        public Configure setLineNumberColor(@ColorInt int lineNumberColor) {
            this.lineNumberColor = lineNumberColor;
            return this;
        }

        public @ColorInt
        int getLineNumberBackColor() {
            return lineNumberBackColor;
        }

        public Configure setLineNumberBackColor(@ColorInt int lineNumberBackColor) {
            this.lineNumberBackColor = lineNumberBackColor;
            return this;
        }
    }
}

File: app/src/main/kotlin/org/codroid/body/widgets/CodroidEditor.kt
package org.codroid.body.widgets

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.util.Log
import android.view.Gravity
import android.view.ViewGroup
import android.widget.PopupWindow
import androidx.annotation.ColorInt
import androidx.appcompat.widget.AppCompatEditText
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.codroid.body.R
import org.codroid.body.dip2px
import org.codroid.interfaces.addon.AddonManager
import org.codroid.interfaces.appearance.AppearanceProperty
import org.codroid.interfaces.appearance.Part
import org.codroid.interfaces.appearance.editor.WrappedSpannable
import org.codroid.interfaces.appearance.parts.EditorPart
import org.codroid.interfaces.evnet.EventCenter
import org.codroid.interfaces.evnet.editor.SelectionChangedEvent
import org.codroid.interfaces.evnet.editor.TextChangedEvent
import java.io.FileReader
import java.io.LineNumberReader
import java.nio.charset.Charset
import java.nio.file.Path
import java.util.*

class CodroidEditor : AppCompatEditText {

    private val mConfigure: Configure by lazy {
        Configure(
            lineNumberEnable = true
        )
    }

    private var mWrappedSpannable: WrappedSpannable
    private var mDelegate: EditorDelegate = EditorDelegate(this)

    private val mLineNumberPaint: Paint by lazy {
        Paint().apply {
            style = Paint.Style.FILL
            textSize = this@CodroidEditor.textSize
            color = mConfigure.lineNumberColor
        }
    }

    private val mBackPaint: Paint by lazy {
        Paint().apply {
            style = Paint.Style.FILL
            color = mConfigure.lineNumberBackColor
        }
    }

    private val lineNumberBackRect: Rect by lazy {
        Rect()
    }

    // This is suggestion window
    private val mPopupWindow: PopupWindow by lazy {
        PopupWindow(
            inflate(context, R.layout.suggestion_window, null),
            ViewGroup.LayoutParams.MATCH_PARENT,
            300
        )
    }

    // The part of editor that defines the appearances
    private var mPart: Optional<Part> = AddonManager.get().appearancePart(AppearanceProperty.PartEnum.EDITOR)

    private var mPaddingLeft = 0F

    // Left margin of main text area if line number is enable.
    private var mGap = 0F

    // return true if editor is drawn at first time.
    // private val isDrawn = false

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {}

    init {
        if (mConfigure.lineNumberEnable) {
            mPart.ifPresent {
                it.findColor(EditorPart.Attribute.BACKGROUND) { it2 ->
                    setBackgroundColor(it2.toArgb())
                }
            }
            textSize = context.dip2px(4F)
        }
        mWrappedSpannable = WrappedSpannable(text)
    }

//    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
//        MeasureSpec.getSize(widthMeasureSpec).let {
//            super.onMeasure(
//                MeasureSpec.makeMeasureSpec(it, MeasureSpec.UNSPECIFIED),
//                heightMeasureSpec
//            )
//        }
//    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
//        super.onLayout(changed, 20, top, 100, bottom)
        if (mConfigure.lineNumberEnable) {
            mGap = context.dip2px(4F)
            mPaddingLeft = mLineNumberPaint.measureText(lineCount.toString()) + mGap * 2
            setPadding((mGap + mPaddingLeft).toInt(), 0, 0, 0)
        }
    }

//    override fun onTouchEvent(event: MotionEvent?): Boolean {
//        if (event?.action == MotionEvent.ACTION_MOVE) return false
//        return super.onTouchEvent(event)
//    }
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (mConfigure.lineNumberEnable) {
            lineNumberBackRect.set(0, 0, mPaddingLeft.toInt(), MeasureSpec.getSize(measuredHeight))
            canvas.drawRect(lineNumberBackRect, mBackPaint)
            for (i in 0 until lineCount) {
                val baseLine = getLineBounds(i, null)
                canvas.drawText(
                    (i + 1).toString(),
                    mGap,
                    baseLine.toFloat(),
                    mLineNumberPaint
                )
            }
        }
    }

    override fun onSelectionChanged(selStart: Int, selEnd: Int) {
        super.onSelectionChanged(selStart, selEnd)
        AddonManager.get().eventCenter()
            .execute<SelectionChangedEvent>(EventCenter.EventsEnum.EDITOR_SELECTION_CHANGED)
            .forEach {
                try {
                    it.onSelectionChanged(mWrappedSpannable.update(text), selStart, selEnd)
                } catch (e: Exception) {
                    AddonManager.get().logger.e("Event editor_selection_changed executes failed with: $e")
                }
            }
    }

    override fun onTextChanged(
        text: CharSequence?,
        start: Int,
        lengthBefore: Int,
        lengthAfter: Int
    ) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter)
        AddonManager.get().eventCenter()
            .execute<TextChangedEvent>(EventCenter.EventsEnum.EDITOR_TEXT_CHANGED)
            .forEach {
                try {
                    it.onTextChanged(
                        mWrappedSpannable.update(getText()),
                        start,
                        lengthBefore,
                        lengthAfter
                    )
                } catch (e: Exception) {
                    AddonManager.get().logger.e("Event editor_text_changed executes failed with: $e")
                }
            }
    }

    fun showSuggestionWindow() {
        mPopupWindow.dismiss()
        mPopupWindow.showAtLocation(
            this, Gravity.TOP or Gravity.START, 0,
            getLineBounds(getCurrentCursorLine() - 1, null) + mPopupWindow.height + lineHeight + 30
        )
    }

    suspend fun fileInput(path: Path, charset: Charset) {
        setText(streamToString(path))
    }

    private suspend fun streamToString(path: Path): String {
        path.toFile().let {
            val buffer = StringBuilder()
            withContext(Dispatchers.IO) {
                LineNumberReader(FileReader(path.toFile())).apply {
                    var i = 0
                    lines().forEach { line ->
                        buffer.append(line).append("\n")
                        if (i > 50) {
                            return@forEach
                        }
                        i++
                        Log.i("Zac", this.lineNumber.toString())
                    }
                }
            }
            return buffer.toString()
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        mDelegate.recycle()
    }

    private fun getCurrentCursorLine(): Int {
        val selectionStart = selectionStart
        return if (selectionStart != -1) {
            layout.getLineForOffset(selectionStart) + 1
        } else -1
    }

    data class Configure(
        val lineNumberEnable: Boolean,
        @ColorInt val lineNumberColor: Int = Color.BLACK,
        @ColorInt val lineNumberBackColor: Int = Color.GRAY,
    )
}
